preprocess/cache_main.cc

```diff
@@ -50,14 +60,65 @@ struct QueueEntry {
   StringPiece *value;
 };
 
-void Input(util::UnboundedSingleQueue<QueueEntry> &queue, util::scoped_fd &process_input, std::unordered_map<uint64_t, StringPiece> &cache, std::size_t flush_rate) {
+void Input(util::UnboundedSingleQueue<QueueEntry> &queue, util::scoped_fd &process_input, std::unordered_map<uint64_t, StringPiece> &cache, std::size_t flush_rate, Options &options) {
   QueueEntry q_entry;
   {
     util::FakeOFStream process(process_input.get());
     std::pair<uint64_t, StringPiece> entry;
     std::size_t flush_count = flush_rate;
+    // Parse column numbers, if given using --key option, into an integer vector (comma separated integers)
+    std::vector<int> columns_args;
+
+    size_t pos_args = 0;
+    std::string token_args;
+    std::string s_args = options.key;
+    while ((pos_args = s_args.find(",")) != std::string::npos) {
+      token_args = s_args.substr(0, pos_args);
+      std::string::size_type sz;
+      int int_token_args = std::stoi (token_args, &sz);
+      columns_args.push_back(int_token_args);
+      s_args.erase(0, pos_args + 1);
+    }
+    std::string::size_type sz;
+    int last_int_token_args = std::stoi (s_args, &sz);
+    columns_args.push_back(last_int_token_args);
+
     for (StringPiece l : util::FilePiece(STDIN_FILENO)) {
-      entry.first = util::MurmurHashNative(l.data(), l.size());
+      // Split content columns, if given using --key option, into a string vector
+      std::vector<std::string> columns;
+ 
+      size_t pos = 0;
+      std::string token;
+      std::string s (l.data(), l.size());
+      while ((pos = s.find(options.field_separator)) != std::string::npos) {
+        token = s.substr(0, pos);
+        columns.push_back(token);
+        s.erase(0, pos + options.field_separator.length());
+      }
+      columns.push_back(s);
+      // Get max/min argument --key to manage some errors
+      int min = std::numeric_limits<int>::max();
+      int max = 0;
+      for (int i : columns_args){
+        if (i < min){
+          min = i;
+        }
+        if (i > max){
+          max = i;
+        }
+      }
+      // If any column number is out of boundaries (or --key is not provided, so uses default -1), use the whole line as data for mmh
+      if (columns.size() < max || min <= 0){
+        entry.first = util::MurmurHashNative(l.data(), l.size());
+      }
+      else{
+        std::string hash_string;
+        for (int i : columns_args){
+          hash_string += columns.at(i-1);
+        }
+        entry.first = util::MurmurHashNative(hash_string.c_str(), hash_string.size());
```

Comment:
I think this chunk can be replaced by `RangeFields` or `IndividualFields` in `fields.hh`